### ISSUE TITLE
setup.py install: Install a "commit-sha" field into coalib

### ISSUE THREAD

**Issue by Makman2** [coala/coala#3367]
As self-installs from a git clone have `dev9999999...` version, we should add a commit sha into the module, like `coalib.__sha__` or `coalib.__commit__` or whatever

This makes bisecting problems easier.

I would also add to the commit string whether the branch was dirty on install, or not, so it looks like
`ab87df1-dirty` or `ab87df1*`

**Comment by Techievena** [comment]
@Makman2 see this [discussion](https://github.com/Homebrew/brew/issues/1738)

**Comment by Makman2** [comment]
@Techievena where's the relevant information?

**Comment by Techievena** [comment]
https://github.com/Homebrew/brew/issues/1738

**Comment by Makman2** [comment]
Your issue is about installing from commit hash, this issue is about placing a commit-sha-variable into our `coalib` module or some other suited module. I'll rephrase the title a bit^^

**Comment by arush0311** [comment]
Can I take this?

**Comment by Makman2** [comment]
@arush0311 how do you want to call the field? :)

**Comment by arush0311** [comment]
There is already a Version in coalib directory, we can add a `__commit__` file when setup.py runs?

**Comment by arush0311** [comment]
@Makman2 but it should only be applicable for install right? ( not develop and not for pip installs ), perhaps will have to read more about setup.py

**Comment by Makman2** [comment]
also for develop and especially for pip installs we should have a `__commit__` :)
you should look at how we set `__version__` :+1:

do you want to support also dirty git branches? I think it's good to have, if someone submits a bug report and you see something like `8afbc21*` then you know that he has untracked modifications made to the repo^^

**Comment by arush0311** [comment]
So the __commit__ will change as soon as there are changes in the repo (in case of develop), perhaps git hooks can help (but they will not be persistent).
And how do we set version, I see there is a version file but who updates that?

**Comment by Makman2** [comment]
no git hooks :) the `__commit__` variable will be generated by `setup.py install` or similar targets. If it detects during installation that it's inside a git repo (or we just require it), it will create a `__commit__` field with the commit sha and checks whether it's a dirty branch or not and modifies/generates `__commit__` accordingly.

[adjust_version_number.py](https://github.com/coala/coala/blob/master/.misc/adjust_version_number.py) is run inside [release.sh](https://github.com/coala/coala/blob/master/.misc/release.sh) which creates the `VERSION` file :+1:

**Comment by arush0311** [comment]
but what if the repo is changed after the installation? in case of develop it will report the wrong `__commit__`

**Comment by Makman2** [comment]
oh yeah `develop`... maybe we put `__commit__ = 'develop'`?

**Comment by arush0311** [comment]
Here are some issues :
1) A python package for checking `git status` will be added if we want to check dirty commits. (current commit can be identified without it.).
2) I couldn't find a way to check for setup.py arguments like "develop", can you help me here? Also discussing with @sils , it wont take care of pip installs. An alternative could be to check for git repo everytime coala runs and update `__commit__` accordingly

**Comment by Makman2** [comment]
1. You could use invoke `git status --porcelain` and check whether it's empty (not dirty) or not (dirty), then you don't need another package.
2. Don't know how to catch `develop` either... But for pip installs you would alter the release script to put already a `__commit__` into `coalib` (like it's done for `__version__` currently).

**Comment by arush0311** [comment]
1) @Makman2 I would still need to make sure that git is installed. (just saying there might be a slight chance that may not be the case)
2) what do you think of the alternative approach for develop ^^

**Comment by Makman2** [comment]
> what do you think of the alternative approach for develop ^^

don't know one :)

**Comment by arush0311** [comment]
> An alternative could be to check for git repo everytime coala runs and update __commit__ accordingly

I was talking about this  :D

**Comment by Makman2** [comment]
you want to update `__commit__` when coala is run? Sounds not right :O

**Comment by arush0311** [comment]
yeah! With this we can get the exact commit someone is getting an error at, even in case of develop.
(I am not proposing to replace updating commit during installation but add the above logic)

**Comment by Makman2** [comment]
But if I've installed coala globally and run it on my own repo? This kind of doesn't work, does it?

**Comment by arush0311** [comment]
Well, that's why there is installation logic as well. (Can you explain ^, I may have misunderstood you 😓 )

**Comment by Makman2** [comment]
I mean the `__commit__` should be set by installation only. Because when I have my own project and want to run coala over it, I don't expect it to suddenly set `coalib.__commit__`, that doesn't make sense.

**Comment by arush0311** [comment]
Well, the global installation will not have a .git dir, even if there is then `__commit__` will only be updated to latest commit of coala? what's wrong with that?

**Comment by arush0311** [comment]
@Makman2 https://github.com/arush0311/coala/commit/ff3cd38269423acb4f63a73be261a45de34b9b1a is the direction i wanted to take. Please take a look ^

**Comment by Makman2** [comment]
direction looks good :+1:

**Comment by Mixih** [comment]
isn't using the name __commit__ a bad idea as it's not a documented magic field?

**Comment by Makman2** [comment]
not sure, isn't `__version__` also undocumented? @Mixih

**Comment by arush0311** [comment]
@Makman2 I think it is https://gitter.im/coala/coala?at=5882d429074f7be763f9d36b

**Comment by Mixih** [comment]
Conclusion is that this aspect is fine, however, @jayvdb may have input

**Comment by jayvdb** [comment]
Have we looked at other tools which may offer this:
- https://github.com/pypa/setuptools_scm
- https://pypi.python.org/pypi/pbr
- https://github.com/warner/python-versioneer
- https://github.com/hawkowl/incremental

**Comment by arush0311** [comment]
Sorry just forgot about this. @jayvdb from what I have seen they automatically set version from git tags + commits, is this what we want to store in out commit variable?
Even then I think we will need to change source code to bend them to our needs
I will look at python-versioneer library, it seems the most customizable (but it does much more than what we ask for)

**Comment by jayvdb** [comment]
I think we want to retain versions with a build date, and these libraries/tools usually support that.

**Comment by Mixih** [comment]
I don't think we want a dated version in the __commit__. We have the version file for that. For __commit, don't we want a commit sha?

**Comment by Mixih** [comment]
@jayvdb @arush0311 so I thought the original goal was to include commit info into a special commit module level attribute for debugging purposes. However, over the course fo the PR, it seems to have changed into including the commit as a part of the version string. Is this still what was requested @Makman2? It seems that including the commit as a part of the version will take more work for the bisection.

**Comment by Makman2** [comment]
The intention was to get the commit from the coala module somehow, so you know to which commit a version belongs and you can easily find the commit on GitHub. Whether this is a separate commit field like `__commit__` or a modified version string doesn't really matter imo. As long as you can extract the commit information somehow without much effort, it's​ fine.
Bisecting with git works with git SHAs, so just the version we currently have doesn't help that well, as you have to match that with the according SHA somehow^^

**Comment by jayvdb** [comment]
As I recall, the goal was that the page on pypi showed the git hash for pre-releases, in addition to the date.  Currently we need to guess the git has based on the date, or pick a random date that we know includes the git commits that we know are needed by the bear.

**Comment by jayvdb** [comment]
Both `pbr` and `incremental` require using `setup_requires`, which has had problems.  I am not sure how reliable that is now, but OpenStack uses `pbr` so it cant be too bad.

`python-versioneer` takes a different approach, of including a [generated](https://github.com/warner/python-versioneer/blob/760c176ab2cc9df9298f344f4c5215bd2039a1c4/src/installer.py#L35) `versioneer.py` at the top level of the source tree.  That means `setup_requires` isnt required as the external dependency is in the source tree.  This is what it looks like : https://gitlab.com/coala/package_manager/merge_requests/95 (I havent looked at the logs yet)

**Comment by Makman2** [comment]
>As I recall, the goal was that the page on pypi showed the git hash for pre-releases, in addition to the date. 

Was it? I just thought of debugging purposes. Actually displaying it inside the pypi README could be another issue imo^^
